<a href="https://colab.research.google.com/github/9bishal/content-based-movie-recommender/blob/main/content_based_movie_recommender.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [460]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))


In [461]:
movies = pd.read_csv('/content/tmdb_5000_movies.csv')
credits = pd.read_csv('/content/tmdb_5000_credits.csv')

In [462]:
movies.head(2)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500

In [463]:
movies.shape

(4803, 20)

In [464]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [465]:
movies = movies.merge(credits,on='title')

In [466]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [467]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [468]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [469]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [470]:
import ast

In [471]:
def convert(text):
    L = []
    for i in ast.literal_eval(text):
        L.append(i['name'])
    return L

In [472]:
movies.dropna(inplace=True)

In [473]:
movies['genres'] = movies['genres'].apply(convert)
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [474]:
movies['keywords'] = movies['keywords'].apply(convert)
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [475]:
import ast
ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [476]:
def convert3(text):
    L = []
    counter = 0
    for i in ast.literal_eval(text):
        if counter < 3:
            L.append(i['name'])
        counter+=1
    return L

In [477]:
movies['cast'] = movies['cast'].apply(convert)
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...   
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...   
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...   
3  [Christian Bale, Michael Caine, Gary Oldman, A...   
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [478]:
movies['cast'] = movies['cast'].apply(lambda x:x[0:3])

In [479]:
def fetch_director(text):
    L = []
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            L.append(i['name'])
    return L

In [480]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [481]:
movies.sample(5)

movie_id                            title  \
2077      8456                  Never Back Down   
2374     27380                        Megaforce   
1772    303858                    Money Monster   
2360      2575             The Tailor of Panama   
3724       510  One Flew Over the Cuckoo's Nest   

                                               overview  \
2077  Rebellious Jake Tyler is lured into an ultimat...   
2374  Megaforce is an elite multi-national military ...   
1772  Financial TV host Lee Gates and his producer P...   
2360  A British spy is banished to Panama after havi...   
3724  While serving time for insanity at a state men...   

                                    genres  \
2077                       [Drama, Action]   
2374  [Adventure, Action, Science Fiction]   
1772                            [Thriller]   
2360                     [Drama, Thriller]   
3724                               [Drama]   

                                               keywords  \
2077  [rebel, martial arts, underground, fight, trai...   
2374  [dictator, soldier, electronic music score, mi...   
1772  [bomb, sniper, tv show, hostage drama, police,...   
2360  [dancing, spy, birthday, map, fireworks, bridg...   
3724  [individual, rebel, self-destruction, wheelcha...   

                                                   cast            crew  
2077            [Sean Faris, Amber Heard, Cam Gigandet]   [Jeff Wadlow]  
2374   [Barry Bostwick, Michael Beck, Persis Khambatta]   [Hal Needham]  
1772    [George Clooney, Julia Roberts, Jack O'Connell]  [Jodie Foster]  
2360  [Pierce Brosnan, Geoffrey Rush, Jamie Lee Curtis]  [John Boorman]  
3724    [Jack Nicholson, Louise Fletcher, Danny DeVito]  [Miloš Forman]

In [482]:
def collapse(L):
    L1 = []
    for i in L:
        L1.append(i.replace(" ",""))
    return L1

In [483]:
movies['cast'] = movies['cast'].apply(collapse)
movies['crew'] = movies['crew'].apply(collapse)
movies['genres'] = movies['genres'].apply(collapse)
movies['keywords'] = movies['keywords'].apply(collapse)

In [484]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

In [485]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [486]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']


In [487]:
new = movies.drop(columns=['overview','genres','keywords','cast','crew'])
new.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [488]:
new['tags'] = new['tags'].apply(lambda x: " ".join(x))
new.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [489]:
new_df = movies[['movie_id', 'title', 'tags']]

In [490]:
new_df['tags']=new_df['tags'].apply(lambda x: " ". join(x))

<ipython-input-490-3aff66d12fc3>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']=new_df['tags'].apply(lambda x: " ". join(x))


In [491]:
new_df['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [492]:
new_df['tags']=new_df['tags'].apply( lambda x: x.lower())

<ipython-input-492-7d4629469e6a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']=new_df['tags'].apply( lambda x: x.lower())


In [493]:
new_df['tags'][0]

'in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron'

In [494]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

##Breaking down the Algorithm

In [495]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000, stop_words='english')

Yes! You got it almost right! Let me refine it a bit for **perfect understanding**. 🚀  

---

## **How `CountVectorizer(max_features=5000, stop_words='english')` Works**
1. **Removes stop words** (like "in", "the", "is", "I", etc.).
2. **Builds a vocabulary** of at most **5000 unique words** (excluding stop words).
3. **Converts text into a numerical representation** (Bag-of-Words model), where:
   - Each row represents a document.
   - Each column represents a word from the vocabulary.
   - The values represent **word frequency** in that document.

---

### **Step-by-Step Example**
#### **Input Texts**
```python
documents = [
    "I love Python and Python loves me!",
    "Python is great for data science.",
    "Machine learning is amazing and useful."
]
```

#### **Step 1: Tokenization (Breaking Text into Words)**
```
["I", "love", "Python", "and", "Python", "loves", "me"]
["Python", "is", "great", "for", "data", "science"]
["Machine", "learning", "is", "amazing", "and", "useful"]
```

---

#### **Step 2: Removing Stop Words (`stop_words='english'`)**
Stop words like **"I", "is", "and", "for"** are removed:
```
["love", "Python", "Python", "loves", "me"]
["Python", "great", "data", "science"]
["Machine", "learning", "amazing", "useful"]
```

---

#### **Step 3: Building a Vocabulary (with `max_features=5000`)**
- Unique words in the dataset (after removing stop words):
  ```
  ['amazing', 'data', 'great', 'learning', 'love', 'loves', 'machine', 'me', 'python', 'science', 'useful']
  ```
- Each word is given an index:
  ```
  amazing → 0
  data → 1
  great → 2
  learning → 3
  love → 4
  loves → 5
  machine → 6
  me → 7
  python → 8
  science → 9
  useful → 10
  ```

---

#### **Step 4: Creating the Bag-of-Words Matrix**
Each document is converted into a numerical **word frequency matrix**:

|  | amazing | data | great | learning | love | loves | machine | me | python | science | useful |
|---|---------|------|-------|----------|------|-------|---------|----|--------|---------|--------|
| **Doc 1** | 0 | 0 | 0 | 0 | 1 | 1 | 0 | 1 | 2 | 0 | 0 |
| **Doc 2** | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 0 | 1 | 1 | 0 |
| **Doc 3** | 1 | 0 | 0 | 1 | 0 | 0 | 1 | 0 | 0 | 0 | 1 |

---

#### **Step 5: Converting to Array**
```python
cv = CountVectorizer(max_features=5000, stop_words='english')
X = cv.fit_transform(documents).toarray()
print(X)
```
Output:
```
[[0 0 0 0 1 1 0 1 2 0 0]
 [0 1 1 0 0 0 0 0 1 1 0]
 [1 0 0 1 0 0 1 0 0 0 1]]
```

---

### **🚀 Final Understanding**
✔ **Stop words removed first.**  
✔ **Vocabulary (max 5000 words) is built.**  
✔ **Each document is converted into a word frequency vector.**  

💡 Now, do you want to test it with your own sentences? 😃

In [496]:
cv.fit_transform(new_df['tags']).toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [497]:
cv.fit_transform(new_df['tags']).toarray().shape

(4806, 5000)

It mean the 4806 movie, and 5000 as the words

In [498]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [499]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [500]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0])

In [501]:
words=cv.get_feature_names_out()

In [502]:
words

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      dtype=object)

In [503]:
import nltk

In [504]:
pip install nltk

##Working with the NPL library

The above data 'word' include some of the unnecessary data like 'dancing' and 'dance'. In order to remove this one we use NPL Library called as `nltk`
So, Basically it works with the `root words`

In [505]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer

In [506]:
def stem(text):
  y=[]
  for i in text.split():
    y.append(ps.stem(i))

  return " ".join(y) #back convert to string

In [507]:
from nltk.stem import PorterStemmer

ps = PorterStemmer()  # Initialize the stemmer

print(ps.stem("loving"))  # Correct way to use stem()


love


Example of the movie data

In [508]:
stem('in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron')

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

Applying to the data 'tags'

In [509]:
new_df['tags']=new_df['tags'].apply(stem)

<ipython-input-509-402dd7cb6b10>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']=new_df['tags'].apply(stem)


In [510]:
from sklearn.metrics.pairwise import cosine_similarity

### **1️⃣ What is Cosine Similarity?**  
Cosine similarity measures how similar two vectors are, **based on their direction rather than their magnitude**.  

It calculates the **cosine of the angle** between two vectors.  

#### **Formula:**
\[
\text{cosine similarity} = \frac{A \cdot B}{||A|| \times ||B||}
\]
where:  
- \( A \cdot B \) is the **dot product** of the two vectors.  
- \( ||A|| \) and \( ||B|| \) are the **magnitudes (norms)** of the vectors.  

#### **Range of values:**
- `1` → **Completely similar** (same direction)  
- `0` → **No similarity** (orthogonal vectors)  
- `-1` → **Completely opposite** (opposite direction)  

---

### **2️⃣ How Does `cosine_similarity(vectors)` Work?**  
- It takes an **array of vectors** (e.g., text embeddings, word frequency counts).  
- Computes the **cosine similarity** between each pair of vectors.  
- Returns a **similarity matrix**.

In [511]:
cosine_similarity(vectors)

array([[1.        , 0.08964215, 0.06071767, ..., 0.02519763, 0.0277885 ,
        0.        ],
       [0.08964215, 1.        , 0.06350006, ..., 0.02635231, 0.        ,
        0.        ],
       [0.06071767, 0.06350006, 1.        , ..., 0.02677398, 0.        ,
        0.        ],
       ...,
       [0.02519763, 0.02635231, 0.02677398, ..., 1.        , 0.07352146,
        0.04774099],
       [0.0277885 , 0.        , 0.        , ..., 0.07352146, 1.        ,
        0.05264981],
       [0.        , 0.        , 0.        , ..., 0.04774099, 0.05264981,
        1.        ]])

In [512]:
cosine_similarity(vectors).shape

(4806, 4806)

In [513]:
similarity=cosine_similarity(vectors)

In [514]:
similarity

array([[1.        , 0.08964215, 0.06071767, ..., 0.02519763, 0.0277885 ,
        0.        ],
       [0.08964215, 1.        , 0.06350006, ..., 0.02635231, 0.        ,
        0.        ],
       [0.06071767, 0.06350006, 1.        , ..., 0.02677398, 0.        ,
        0.        ],
       ...,
       [0.02519763, 0.02635231, 0.02677398, ..., 1.        , 0.07352146,
        0.04774099],
       [0.0277885 , 0.        , 0.        , ..., 0.07352146, 1.        ,
        0.05264981],
       [0.        , 0.        , 0.        , ..., 0.04774099, 0.05264981,
        1.        ]])

In [515]:
similarity[0]

array([1.        , 0.08964215, 0.06071767, ..., 0.02519763, 0.0277885 ,
       0.        ])

This mean, the 1st movies have this much of similarrity to every movie

In [516]:
similarity[0].shape

(4806,)

In [517]:
similarity[0]

array([1.        , 0.08964215, 0.06071767, ..., 0.02519763, 0.0277885 ,
       0.        ])

(0)th represents, the first movie is simililar to itsellf

In [518]:
similarity[1]

array([0.08964215, 1.        , 0.06350006, ..., 0.02635231, 0.        ,
       0.        ])


(1)st, the second movie is simililar to itsellf

We just neewd to find the 5 movies that are similar to the currently watching movie, using the `sorted()` function we could easily find it,

In [519]:
sorted(similarity[0], reverse=True)

[1.0,
 0.26089696604360174,
 0.2581988897471611,
 0.25110592822973776,
 0.24944382578492943,
 0.24846467329894412,
 0.24397501823713333,
 0.243599382882345,
 0.24147264420814754,
 0.23904572186687872,
 0.22677868380553634,
 0.2238868314198225,
 0.22230800575069137,
 0.22190115272469205,
 0.22131333406899523,
 0.2173253797873328,
 0.21653278478430665,
 0.2164218276749025,
 0.2162249910469341,
 0.2162249910469341,
 0.2138089935299395,
 0.2125118592516207,
 0.21251185925162067,
 0.21251185925162067,
 0.2114722130550724,
 0.21147221305507238,
 0.2059714602177749,
 0.20498001542269692,
 0.20283702113484398,
 0.2,
 0.1980534816610477,
 0.19518001458970666,
 0.19518001458970666,
 0.19518001458970663,
 0.19518001458970663,
 0.19477964490741226,
 0.19451950503185494,
 0.1938916835823703,
 0.1938916835823703,
 0.19320245583349133,
 0.19194297398747862,
 0.1916629694999819,
 0.18869127060994534,
 0.18832944617230335,
 0.18752289237539818,
 0.18752289237539815,
 0.18708286933869708,
 0.18442777839

We need to keep even the indexes, we `enumerated()` function is used, and list is used to convert into list

It mean that the, the movie at the 0th index is "0.08964214570007951" distance from the 1st index movie

In [520]:
list(enumerate(similarity[0]))

[(0, 1.0),
 (1, 0.08964214570007951),
 (2, 0.060717674071891624),
 (3, 0.03984095364447979),
 (4, 0.18257418583505536),
 (5, 0.1183452670827877),
 (6, 0.023218173010628604),
 (7, 0.1763834207376394),
 (8, 0.06388765649999398),
 (9, 0.07733088933397005),
 (10, 0.11769797726729994),
 (11, 0.081199794294115),
 (12, 0.09583148474999097),
 (13, 0.04733810683311508),
 (14, 0.11428571428571427),
 (15, 0.05096471914376255),
 (16, 0.08226127456606226),
 (17, 0.15026857675938213),
 (18, 0.11194341570991125),
 (19, 0.08728715609439694),
 (20, 0.058167505074711086),
 (21, 0.09416472308615166),
 (22, 0.06900655593423542),
 (23, 0.09759000729485331),
 (24, 0.0563436169819011),
 (25, 0.03877833671647406),
 (26, 0.17142857142857143),
 (27, 0.18752289237539818),
 (28, 0.12555296411486888),
 (29, 0.07100716024967263),
 (30, 0.07319250547113999),
 (31, 0.15971914124998493),
 (32, 0.09116846116771035),
 (33, 0.1057361065275362),
 (34, 0.0),
 (35, 0.11065666703449761),
 (36, 0.19194297398747862),
 (37, 0.0

Here, we used lamda function to clearify that, sorting by the not index, but by the 2nd number like "0.26089696604360174"

In [521]:
sorted(list(enumerate(similarity[0])), reverse=True, key=lambda x:x[1])

[(0, 1.0),
 (539, 0.26089696604360174),
 (1192, 0.2581988897471611),
 (260, 0.25110592822973776),
 (1214, 0.24944382578492943),
 (507, 0.24846467329894412),
 (1440, 0.24397501823713333),
 (1916, 0.243599382882345),
 (582, 0.24147264420814754),
 (3728, 0.23904572186687872),
 (74, 0.22677868380553634),
 (3606, 0.2238868314198225),
 (322, 0.22230800575069137),
 (83, 0.22190115272469205),
 (4190, 0.22131333406899523),
 (2329, 0.2173253797873328),
 (1202, 0.21653278478430665),
 (47, 0.2164218276749025),
 (61, 0.2162249910469341),
 (495, 0.2162249910469341),
 (2405, 0.2138089935299395),
 (1533, 0.2125118592516207),
 (942, 0.21251185925162067),
 (972, 0.21251185925162067),
 (466, 0.2114722130550724),
 (1327, 0.21147221305507238),
 (2200, 0.2059714602177749),
 (151, 0.20498001542269692),
 (1199, 0.20283702113484398),
 (973, 0.2),
 (4046, 0.1980534816610477),
 (3160, 0.19518001458970666),
 (4405, 0.19518001458970666),
 (1273, 0.19518001458970663),
 (2782, 0.19518001458970663),
 (2969, 0.1947796

But we only neede 5 item

In [522]:
sorted(list(enumerate(similarity[0])), reverse=True, key=lambda x:x[1])[1:6]

[(539, 0.26089696604360174),
 (1192, 0.2581988897471611),
 (260, 0.25110592822973776),
 (1214, 0.24944382578492943),
 (507, 0.24846467329894412)]

Suming up it all, we conlude:

In [523]:
from operator import index
def recommned(movie):
  movie_index = new_df[new_df['title'] == movie].index[0]
  distances = similarity[movie_index]
  movie_list = sorted(list(enumerate(distances)), reverse=True, key=lambda x:x[1])[1:6]

  for i in movie_list:
    print(i[0])

In [524]:
recommned('Avatar')

539
1192
260
1214
507


In [525]:
new_df.iloc[1216]

movie_id                                                10641
title                                      Autumn in New York
tags        autumn in new york follow the sexual exploit o...
Name: 1216, dtype: object

In [526]:
new_df.iloc[1216].title

'Autumn in New York'

In [537]:
from operator import index
def recommned(movie):
  movie_index = new_df[new_df['title'] == movie].index[0]
  distances = similarity[movie_index]
  movie_list = sorted(list(enumerate(distances)), reverse=True, key=lambda x:x[1])[1:6]

  for i in movie_list:
    print(new_df.iloc[i[0]].title)
    # print(i[0])

In [538]:
recommned('Batman Begins')

The Dark Knight
The Dark Knight Rises
Batman
Batman & Robin
Batman
